### PR TITLE
MGMT-2636 Remove uploadBaseISO job

### PR DESCRIPTION
**Pull request by djzager**
This way, on startup, we simply upload the base ISO to s3 using leader
election without the additional burden of running, and waiting for, a
job to run.

### CODE REVIEW

File: pkg/s3wrapper/client.go [hunk 1]
```diff
@@ -32,6 +32,7 @@ const awsEndpointSuffix = ".amazonaws.com"
 
 // TODO: Can we use RHCOS_VERSION or BASE_OS_IMAGE in the Makefile to set the value for
 // BaseObjectName?
+const RHCOSBaseURL = "https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/rhcos-4.6/46.82.202009222340-0/x86_64/rhcos-46.82.202009222340-0-live.x86_64.iso"
```
Comment (carbonin) [inline pkg/s3wrapper/client.go:35]: Where was this value coming from before?
I mean, I can tell it was part of the image build, but should this be in configuration somewhere rather than hardcoded?

Comment (carbonin) [reply pkg/s3wrapper/client.go:35]: Oh it's in the Makefile and now copied here ... Can we deduplicate this somehow?

Comment (djzager) [reply pkg/s3wrapper/client.go:35]: Deduped it by taking it out of the Makefile.

File: docs/installer-live-iso.md [hunk 2]
```diff
@@ -39,14 +39,13 @@ sed -i 's#quay.io/ocpmetal/assisted-service:latest#'"${SERVICE}"'#' onprem-iso-c
...
 
 ````
-podman run --privileged --pull=always --rm \
-  -v .:/download -w /download \
-  --entrypoint /bin/bash \
-  quay.io/ocpmetal/assisted-iso-create:latest cp /data/livecd.iso /download/livecd.iso
+podman run --privileged --pull=always -it --rm \
+  -v .:/data -w /data --entrypoint /bin/bash \
+  quay.io/coreos/coreos-installer:v0.7.0 \
+  -c 'coreos-installer download -f iso --decompress && mv fedora-coreos-*.iso livecd.iso'
```
Comment (eranco74) [inline docs/installer-live-iso.md:48]: fedora-coreos?
Should be rhcos-*-live.x86_64.iso no?

File: cmd/main.go [hunk 3]
```diff
@@ -436,8 +436,11 @@ func uploadBaseISOWithLeader(uploadLeader leader.ElectorInterface, objectHandler
 	}
 	return uploadLeader.RunWithLeader(ctx, func() error {
 		log.Info("Starting base ISO upload")
-		err = generator.UploadBaseISO()
-		log.Info("Finished base ISO upload")
-		return err
+		resp, err := http.Get(s3wrapper.RHCOSBaseURL)
```
Comment (eranco74) [inline cmd/main.go:439]: Why use a closure and not execute the logic here?
It seems that the code calling this function might log an error "Failed uploading base ISO" before the upload logic is executed.
https://sourcegraph.com/github.com/openshift/assisted-service@66ab7e7ad61f528b74fda9299538edabee6439b9/-/blob/cmd/main.go#L350

**Comment by avishayt** [comment]
It looks like we will soon have 2 more image formats to upload at startup: PXE and small ISO that downloads the rootfs.  In this case it's probably better to leave the job.  What do you think?

**Comment by djzager** [comment]
> It looks like we will soon have 2 more image formats to upload at startup: PXE and small ISO that downloads the rootfs. In this case it's probably better to leave the job. What do you think?

Probably. The important question is how we intend to get the set of images to upload to the assisted-service (or the job it spawns)?

**Comment by djzager** [comment]
The other thing that goes into whether or not we keep the separate job...what exactly does it buy us if the service isn't ready until the job is run to completion? From my perspective, if the work is to be done synchronously, then a Job is not a good fit.

**Comment by filanov** [comment]
> The other thing that goes into whether or not we keep the separate job...what exactly does it buy us if the service isn't ready until the job is run to completion? From my perspective, if the work is to be done synchronously, then a Job is not a good fit.

If the job is not done the user will not be able to generate an ISO, we can always return 409, not sure if it's a good UX, in 99% of the time the job will just check if the base image already exist and just exist, after all we don't add new images every release.

**Comment by mhrivnak** [comment]
It seems there's a choice between:
1. process launches
2. process creates a Job that runs some block of code
3. process waits for the Job to finish successfully
4. process deletes the Job (from a function called "monitor" :eyes: )
5. process continues its startup and begins serving its API

or

1. process launches
2. process runs some block of code
3. process retries until block of code succeeds
4. process continues its startup and begins serving its API

I don't see how a Job is helpful. It introduces extra complexity that we can eliminate:
* Interacting with the k8s API and depending on it to schedule the Job
* Just creating the Job itself is 100+ lines of code
* we have to build a separate container image to run the Job? Or use the same image but run a different binary? Or run the same binary with different arguments? This could instead just be a function in our main binary.
* monitoring and deleting the Job (25 lines of code, polling the k8s API) is probably more complex than a little retry logic would be in golang.
* less traffic on the k8s api server is always good

Jobs are useful in a fire-and-forget scenario, a fan-out scenario, or at least a concurrency scenario. We don't have any of those here. I can see how a Job might have been helpful if the workflow had evolved differently, but I don't see it here.

**Comment by filanov** [comment]
> * process launches
> * process runs some block of code
> * process retries until block of code succeeds
> * process continues its startup and begins serving its API

It's a good point, but where do we get he image from and where do we store it? 
in on-prem we will need to have the image somewhere in the file system 
in the cloud we will need to probably download it, so we will have two different base codes for it? 
And what will we do with multiple images?

File: README.md [hunk 4]
```diff
@@ -50,7 +51,8 @@ skipper make generate-from-swagger
 
 ### Pre-configuration
 
-- Run minikube on your system.
+- Run minikube on your system. **NOTE: You must enable the registry addon in
+    minikube with `minikube addons enable registry`.**
```
Comment (sacharya) [inline README.md:55]: Just curious, what's the need for the registry addon?

Comment (djzager) [reply README.md:55]: The `deploy-test` target eventually runs `_update_minikube` that builds images on the minikube host. We need the registry addons for this to work properly.

It's hard to remember what combination of `ASSISTED_ORG=minikube-local-registry` or `minikube docker-env` makes the registry necessary. The only thing I do remember is not being able to run through test scenarios using `make` until I enabled it.

**Review by rwsu — COMMENTED** [review]
Looks good overall.

There are two more places where assisted-iso-create is being referenced:
* https://github.com/openshift/assisted-service/blob/master/tools/deploy_assisted_installer.py#L67
* https://github.com/openshift/assisted-service/blob/master/tools/deploy_assisted_installer_configmap.py#L62

Do these need to be updated?

**Comment by djzager** [comment]
@rwsu I updated the files mentioned. Thanks for pointing them out.

**Comment by avishayt** [comment]
Looks fine to me

File: cmd/main.go [hunk 5]
```diff
@@ -434,7 +436,35 @@ func autoMigrationWithLeader(migrationLeader leader.ElectorInterface, db *gorm.D
...
+		// Based on the ISO 9660 standard we should consistently be able to
+		// grab the Volume ID here.
+		volumeId := make([]byte, 32)
+		_, err = iso.ReadAt(volumeId, 32808)
+		if err != nil {
+			log.WithError(err).Fatal("Failed to read volume id from provided base image")
+			return err
+		}
+		// TODO: Should we verify that the volume id is of the form `rhcos-<version>`?
+		s3wrapper.BaseObjectName = strings.TrimSpace(string(volumeId)) + ".iso"
```
Comment (carbonin) [inline cmd/main.go:464]: Is there a more clear way we could do this?
Maybe by using [this iso package](https://github.com/kdomanski/iso9660/blob/880de9cb649c5731170a2f808b11ef1392954864/image_reader.go#L52)?

Full disclosure: I'm not entirely sure that function is doing the same thing you are here, but I usually look for alternatives when I see something like `_, err = iso.ReadAt(volumeId, 32808)` :sweat_smile:

File: pkg/s3wrapper/client.go [hunk 6]
```diff
@@ -30,9 +30,14 @@ import (
...
-// BaseObjectName?
-const BaseObjectName = "livecd-46.82.202009222340-0.iso"
+// TODO: Eventually we should make the base image piece managed out-of-band.
+// Here, we are providing a fallback in the case an image isn't provided.
+const RHCOSBaseURL = "https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.6/4.6.1/rhcos-4.6.1-x86_64-live.x86_64.iso"
+const RHCOSBaseObjectName = "rhcos-46.82.202010091720-0.iso"
+
+// We will need to modify this based on whatever is provided to the
+// assisted-service at runtime.
+var BaseObjectName string
```
Comment (carbonin) [inline pkg/s3wrapper/client.go:40]: Can this be provided as a part of `NewS3Client` rather than a global? Maybe in the config somewhere?

File: README.md [hunk 7]
```diff
@@ -11,6 +11,7 @@
 1. skipper <https://github.com/stratoscale/skipper>
 1. minikube (for tests)
 1. kubectl
+1. Python modules `pip install waiting`
```
Comment (carbonin) [inline README.md:14]: Why is this needed?

Comment (djzager) [reply README.md:14]: https://github.com/openshift/assisted-service/blob/master/tools/wait_for_assisted_service.py#L3

File: cmd/main.go (hunk 5 above)
Comment (djzager) [reply cmd/main.go:464]: I think I could add a more detailed comment about the bytes that we are skipping and why. But based on my reading of different implementations of iso readers, the location of the 32 bytes for the volume id is pretty fixed. I'm personally hesitant to bring in a whole iso package just for this.

WDYT about my explaining exactly what is happening in a comment.

For reference, this is where the volume id is being set when RHCOS images are built https://github.com/coreos/coreos-assembler/blob/master/src/cmd-buildextend-live#L103

File: cmd/main.go (hunk 3 above)
Comment (djzager) [reply cmd/main.go]: I don't think I addressed this. I'll have another look at this tomorrow.

File: cmd/main.go (hunk 5 above)
Comment (avishayt) [reply cmd/main.go:464]: Don't worry about bringing that package in - I'm already using it in a PR that is based on this one. :)
That said, I'm not sure the library has a method for getting the volume ID.  From my experience with it it's more for reading files/directories inside the ISO, whereas the volume ID is probably in some metadata area.  Might be wrong though.

File: README.md (hunk 4 above)
Comment (filanov) [reply README.md:55]: @YuviGold maybe it can replace our `eval` use?

File: cmd/main.go [hunk 8]
```diff
@@ -434,7 +436,35 @@ func autoMigrationWithLeader(migrationLeader leader.ElectorInterface, db *gorm.D
 }
 
 func uploadBaseISOWithLeader(uploadLeader leader.ElectorInterface, objectHandler s3wrapper.API, generator generator.ISOInstallConfigGenerator, log logrus.FieldLogger) error {
```
Comment (filanov) [inline cmd/main.go:438]: maybe this section should move to a different package? it's a lot of code.

Comment (filanov) [inline cmd/main.go:438]: how does all this code tested? in case someone will make changes.

File: Makefile [hunk 9]
```diff
@@ -264,7 +248,7 @@ deploy-onprem:
 	podman pod create --name assisted-installer -p 5432,8000,8090,8080
 	# These are required because when running on RHCOS livecd, the coreos-installer binary and
 	# livecd are bind-mounted from the host into the assisted-service container at runtime.
-	[ -f livecd.iso ] || curl $(BASE_OS_IMAGE) --retry 5 -o livecd.iso
+	[ -f livecd.iso ] || until curl https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.6/latest/rhcos-live.x86_64.iso --retry 5 -o livecd.iso; do sleep 1; done
```
Comment (avishayt) [inline Makefile:251]: why hard-code the URL?

Comment (YuviGold) [inline Makefile:251]: How we can support different RHCOS images?

File: README.md (hunk 7 above)
Comment (YuviGold) [reply README.md:14]: It is installed on https://github.com/openshift/assisted-service/blob/master/Dockerfile.assisted-service-build#L15
Maybe we should split into a `requirements.txt` file that could be used even without the Dockerfile.
Than inside the Dockerfile we should use
```dockerfile
COPY requirements.txt .
python3 -m pip install -r requirements.txt
```

File: README.md (hunk 4 above)
Comment (YuviGold) [reply README.md:55]: It does. I was checking for this alternative. But I remember I didn't choose it because it requires manual minikube configuration. 

@djzager 
But out of curiosity, why is that better?
You shouldn't call directly to `minikube-local-registry` and `minikube docker-env` while `deploy-test` takes care of all the process.

File: tools/deploy_assisted_installer.py [hunk 10]
```diff
@@ -63,10 +63,10 @@ def main():
             service_container["volumeMounts"].append({'name': 'iso', 'mountPath': "/data"})
             spec["volumes"].append({'name': 'iso', 'emptyDir': {}})
             spec["initContainers"] = [{
-                "name": "assisted-iso-create",
-                "image": "quay.io/ocpmetal/assisted-iso-create:latest",
+                "name": "rhcos-download",
+                "image": "quay.io/app-sre/centos:8",
                 "command": ["bash", "-c"],
-                "args": ["cp -r /data/* /iso-data"],
+                "args": ["until curl https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.6/latest/rhcos-live.x86_64.iso --retry 5 -o /iso-data/livecd.iso; do sleep 1; done"],
```
Comment (YuviGold) [inline tools/deploy_assisted_installer.py:69]: Doesn't this need to be dependent on the `OPENSHIFT_VERSION`?
Like `https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/{OPENSHIFT_VERSION}/latest/rhcos-live.x86_64.iso`

File: cmd/main.go (hunk 8 above)
Comment (avishayt) [reply cmd/main.go:438]: @filanov I agree with you, but I am already almost done with a PR that is based on this, and I am moving the code to s3wrapper together with code from my PR that uploads PXE artifacts.  Suggest we continue with this and my PR will move it.

Comment (filanov) [reply cmd/main.go:438]: No problem but then you will need to write tests for everything :D

File: pkg/s3wrapper/client.go (hunk 6 above)
Comment (djzager) [reply pkg/s3wrapper/client.go:40]: It can be. The reason I had gone with this approach was to avoid updating all the calls to `s3wrapper.BaseObjectName` to `s3wrapper.BaseObjectName()`.

It made me a little uneasy to make it a global like this, let me know if I should make it private and use accessors.

File: cmd/main.go (hunk 8 above)
Comment (avishayt) [reply cmd/main.go:438]: Will add tests for this, sure

File: README.md (hunk 7 above)
Comment (djzager) [reply README.md:14]: I think that is a reasonable approach. For now, I think calling out the need for `waiting` is sufficient for someone to make it through the README.

File: Makefile (hunk 9 above)
Comment (djzager) [reply Makefile:251]: Resolving this in favor of https://github.com/openshift/assisted-service/pull/645#discussion_r538139450

File: cmd/main.go (hunk 5 above)
Comment (djzager) [reply cmd/main.go:464]: Since I don't see anything in that package that, at least in a straightforward way, gives me the volume id I want, I will comment the `ReadAt(volumeId, 32808)` to communicate where `32808` comes from.

File: README.md (hunk 4 above)
Comment (djzager) [reply README.md:55]: I leave open the possibility that I was missing something when I first tried to go through the README, but I wasn't able to get the `make deploy-test` to work until I enabled the minikube registry config option. I'm happy to remove it and investigate it separately. Sound reasonable?

File: cmd/main.go (hunk 8 above)
Comment (djzager) [reply cmd/main.go:438]: I think pushing this into a different package and making it unit-testable is the right approach. To avoid conflicts with @avishayt I'll skip this item, _unless_ addressing this comment (https://github.com/openshift/assisted-service/pull/645#discussion_r523721694) forces my hand.

File: pkg/s3wrapper/client.go (hunk 6 above)
Comment (djzager) [reply pkg/s3wrapper/client.go:40]: I couldn't find an intelligent way to do this that didn't make integrating with https://github.com/openshift/assisted-service/pull/755 more painful for @avishayt so I'm skipping it and will circle back after #755 is merged.

File: tools/deploy_assisted_installer.py (hunk 10 above)
Comment (djzager) [reply tools/deploy_assisted_installer.py]: Since we are falling back to this image if one isn't provided to the assisted-service at runtime, I elected to just remove the initContainer. That prevents us from having to maintain it in more than the 1) Makefile and 2) s3wrapper. WDYT?

File: Makefile (hunk 9 above)
Comment (djzager) [reply Makefile]: Added ability to configure.

File: cmd/main.go (hunk 5 above)
Comment (djzager) [reply cmd/main.go]: Added a comment to explain this. 32768 bytes are unused and volume id is 40 bytes in. Let me know if the comment is sufficient.

https://wiki.osdev.org/ISO_9660
https://en.wikipedia.org/wiki/ISO_9660

File: internal/bminventory/inventory.go [hunk 11]
```diff
@@ -839,7 +837,6 @@ func (b *bareMetalInventory) GenerateClusterISO(ctx context.Context, params inst
 	updates["image_ssh_public_key"] = params.ImageCreateParams.SSHPublicKey
 	updates["image_created_at"] = strfmt.DateTime(now)
 	updates["image_expires_at"] = strfmt.DateTime(now.Add(b.Config.ImageExpirationTime))
-	updates["image_generator_version"] = b.Config.ImageBuilder
```
Comment (carbonin) [inline internal/bminventory/inventory.go:842]: For a followup, but I assume we can eventually remove this column, right?

File: cmd/main.go [hunk 12]
```diff
@@ -434,7 +436,36 @@ func autoMigrationWithLeader(migrationLeader leader.ElectorInterface, db *gorm.D
...
+		if err != nil {
+			log.WithError(err).Fatal("Failed to read volume id from provided base image")
+			return err
+		}
+		// TODO: Should we verify that the volume id is of the form `rhcos-<version>`?
+		s3wrapper.BaseObjectName = strings.TrimSpace(string(volumeId)) + ".iso"
+	}
+
+	// Second, we check if it is already in s3
 	exists, err := objectHandler.DoesObjectExist(ctx, s3wrapper.BaseObjectName)
```
Comment (carbonin) [inline cmd/main.go:469]: Should we move this check into the `RunWithLeader` block?
Otherwise each replica could see this as `false` and then sequentially upload the file multiple times as each acquires the leader lock.

File: internal/bminventory/inventory.go (hunk 11 above)
Comment (filanov) [reply internal/bminventory/inventory.go:842]: from the swagger as well.

**Comment by avishayt** [comment]
@filanov Anything else here or ready to merge?

**Comment by filanov** [comment]
> @filanov Anything else here or ready to merge?

Just want to make sure that test-infra is passing so a jenkins link will be nice

**Comment by filanov** [comment]
and a rebase

**Comment by filanov** [comment]
still waiting for test infra results.

**Comment by YuviGold** [comment]
@djzager Needs rebase.

**Comment by carbonin** [comment]
So we can't actually run test infra properly for this (or any PR) really due to https://github.com/openshift/assisted-test-infra/pull/341#issuecomment-749162708

In some cases it might be fine to just replace the service image, but when we're also editing the deploy scripts we need a clean checkout (it seems) to get test infra to run properly.

**Comment by carbonin** [comment]
That said, I was able to run through a deploy with test infra on my test machine using this branch, but it looks like I'll have to test again after this next rebase.

**Comment by carbonin** [comment]
Tested another cluster install after the rebase :+1: :shipit: 

/unhold
/lgtm

**Review by carbonin — APPROVED** [review]

**Comment by YuviGold** [comment]
> So we can't actually run test infra properly for this (or any PR) really due to https://github.com/openshift/assisted-test-infra/pull/341#issuecomment-749162708
> 
> In some cases it might be fine to just replace the service image, but when we're also editing the deploy scripts we need a clean checkout (it seems) to get test infra to run properly.

You can test the service by providing:
* SERVICE_REPO, SERVICE_BRANCH - for the deployment scripts
* SERVICE - The image going to be used for python client and the deployment itself.

If both were provided and still it didnt test properly, so there might be a bug. Please share a link next time you see that.

**Comment by filanov** [comment]
Great job everyone
